shared/pubsub.js

```diff
@@ -0,0 +1,397 @@
+ * {boolean?} reconnectOnOnline - Whether to reconnect after regaining internet connectivity.
+ * {boolean?} reconnectOnTimeout - Whether to reconnect after a timeout.
+ * @returns {Object}
+ */
+export function createClient (url: string, options?: Object = {}): PubSubClient {
+  const client = {
+    customHandlers: options.handlers || {},
+    failedAttempts: 0,
+    isReconnecting: false,
+    listeners: Object.create(null),
+    messageHandlers: { ...defaultMessageHandlers, ...options.messageHandlers },
+    options: { ...defaultOptions, ...options },
+    pendingSubscriptionSet: new Set(),
+    pendingUnsubscriptionSet: new Set(),
+    reconnectionDelayID: undefined,
+    socket: null,
+    subscriptionSet: new Set(),
+    connectionTimeoutID: undefined,
+    url: url.replace(/^http/, 'ws'),
+    // Methods
+    ...{ connect, destroy, getNextRandomReconnectionDelay, pub, sub, unsub }
+  }
+  // Create and save references to reusable event listeners.
+  for (const name of [...eventNames, ...globalEventNames]) {
+    client.listeners[name] = (event: Event, ...args: any[]) => {
+      if (client.options.debug) {
+        console.debug('[pubsub] Event:', name, ...args)
+      }
+      const customHandler = client.customHandlers[name]
+      const defaultHandler = defaultHandlers[name]
+
+      if (defaultHandler) {
+        defaultHandler.call(client, event)
+      }
+      if (customHandler) {
+        customHandler.call(client, event)
+      }
+    }
+  }
+  // Attach global event listeners before the first connection.
+  if (typeof window === 'object') {
+    globalEventNames.forEach((name) => {
+      window.addEventListener(name, client.listeners[name])
+    })
+  }
+  client.connect()
+  return client
+}
+
+export function createMessage (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+export function createRequest (type: $Values<typeof REQUEST_TYPE>, data: JSONObject): string {
+  // Had to use Object.assign() instead of object spreading to make Flow happy.
+  return JSON.stringify(Object.assign({ type }, data))
+}
+
+// ====== Event Handlers ====== //
+
+const defaultHandlers = {
+  close (event: CloseEvent) {
+    if (this.socket) {
+      // Remove event listeners to avoid memory leaks.
+      for (const name of eventNames) {
+        this.socket.removeEventListener(name, this.listeners[name])
+      }
+    }
+    this.socket = null
+
+    if (this.connectionTimeoutID) {
+      clearTimeout(this.connectionTimeoutID)
+      this.connectionTimeoutID = undefined
+    }
+    if (this.isReconnecting) {
+      this.failedAttempts++
+    }
+    if (event.reason === 'timeout' && this.options.reconnectOnTimeout) {
+      this.reconnectionDelayID = setTimeout(
+        this.connect,
+        this.getNextRandomReconnectionDelay()
+      )
+    }
+  },
+
+  // Do not manually close the socket here as it will be done automatically.
+  error (event: Event) {
+  },
+
+  message (event: MessageEvent) {
```

Comment:
What is this `MessageEvent` type? I don't see where it's defined.

Reply:
It's a standard interface defined in the HTML spec: https://html.spec.whatwg.org/#the-messageevent-interface